### PR TITLE
[Bug] hot fix to enable L3 E2E

### PR DESCRIPTION
**Pull request by er1cthe0ne**
This is the change to enable L3 E2E, with this change, we confirmed the L3 routing E2E is working through the horizon UI.

This PR includes the follow changes:

1. added setup_ovs_bridges_if_need in aca_main which will work nicely when there is no ovs bridge when aca starts
2. allow neighbor state to be create/update/info, since we are getting neighbor state update during L3 E2E
3. enable process of neighbor state with fixed_ips_size > 1
4. fixed the path issue with ovs-ctl

### CODE REVIEW

File: src/dp_abstraction/aca_dataplane_ovs.cpp [hunk 1]
```diff
@@ -211,11 +214,15 @@ int ACA_Dataplane_OVS::update_port_state_workitem(const PortState current_PortSt
     break;
 
   case OperationType::NEIGHBOR_CREATE_UPDATE:
-
+    /* deprecating */
```
Comment (xieus) [inline src/dp_abstraction/aca_dataplane_ovs.cpp:217]: If we can use some annotation in the codes, that would look more formal :-)

One example in Java: https://github.com/futurewei-cloud/alcor/blob/0276c273e91de79e6b603687810afa730e5a4117/lib/src/main/java/com/futurewei/alcor/common/repo/ICache.java#L23

File: src/dp_abstraction/aca_dataplane_ovs.cpp [hunk 2]
```diff
@@ -121,8 +121,11 @@ int ACA_Dataplane_OVS::update_port_state_workitem(const PortState current_PortSt
...
 
     try {
-      // TODO: add support for more than one fixed_ips in the future
-      assert(current_PortConfiguration.fixed_ips_size() == 1);
+      if (current_PortConfiguration.fixed_ips_size() <= 0) {
+        ACA_LOG_ERROR("PortConfiguration.fixed_ips_size: %d.\n",
+                      current_PortConfiguration.fixed_ips_size());
+        throw std::invalid_argument("PortConfiguration.fixed_ips_size is less than zero");
+      }
       virtual_ip_address = current_PortConfiguration.fixed_ips(0).ip_address();
```
Comment (xieus) [inline src/dp_abstraction/aca_dataplane_ovs.cpp:129]: Although we allow more than fixed IPs, we process the first one only as the primary IP.  As a next step, it would be awesome to start thinking about storing the secondary IPs so that when customers add IPs in the VM (manually as limited by current networking stack), Control Plane has the knowledge to grant or reject customers' traffic, depending on whether customers configure the VM in the right way.

File: src/dp_abstraction/aca_dataplane_ovs.cpp [hunk 3]
```diff
@@ -354,121 +361,138 @@ int ACA_Dataplane_OVS::update_neighbor_state_workitem(NeighborState current_Neig
...
+  case OperationType::UPDATE:
+    [[fallthrough]];
+  case OperationType::INFO:
 
-    // TODO: add support for more than one fixed_ips in the future
-    assert(current_NeighborConfiguration.fixed_ips_size() == 1);
+    if (current_NeighborConfiguration.fixed_ips_size() <= 0) {
+      ACA_LOG_ERROR("NeighborConfiguration.fixed_ips_size: %d.\n",
+                    current_NeighborConfiguration.fixed_ips_size());
+      throw std::invalid_argument("NeighborConfiguration.fixed_ips_size is less than zero");
```
Comment (xieus) [inline src/dp_abstraction/aca_dataplane_ovs.cpp:372]: This validation is in use more than one times. Let us keep the logic in a method if possible.

File: src/dp_abstraction/aca_dataplane_ovs.cpp [hunk 4]
```diff
@@ -354,121 +361,138 @@ int ACA_Dataplane_OVS::update_neighbor_state_workitem(NeighborState current_Neig
...
+    }
 
-    if (current_NeighborConfiguration.fixed_ips(0).neighbor_type() == NeighborType::L2 ||
-        current_NeighborConfiguration.fixed_ips(0).neighbor_type() == NeighborType::L3) {
-      try {
-        assert(current_NeighborConfiguration.fixed_ips_size() == 1);
-        virtual_ip_address = current_NeighborConfiguration.fixed_ips(0).ip_address();
+    for (int ip_index = 0;
+         ip_index < current_NeighborConfiguration.fixed_ips_size(); ip_index++) {
+      if (current_NeighborConfiguration.fixed_ips(ip_index).neighbor_type() == NeighborType::L2 ||
```
Comment (xieus) [inline src/dp_abstraction/aca_dataplane_ovs.cpp:377]: As we use current_NeighborConfiguration.fixed_ips(ip_index) for more than once, recommend to use a local variable to store it.

File: src/dp_abstraction/aca_dataplane_ovs.cpp [hunk 5]
```diff
@@ -354,121 +361,138 @@ int ACA_Dataplane_OVS::update_neighbor_state_workitem(NeighborState current_Neig
...
-                NeighborType::L3) {
-              overall_rc = ACA_OVS_L3_Programmer::get_instance().create_neighbor_l3(
-                      current_NeighborConfiguration.vpc_id(),
-                      current_NeighborConfiguration.fixed_ips(0).subnet_id(), found_network_type,
-                      virtual_ip_address, virtual_mac_address, host_ip_address,
+            ACA_LOG_DEBUG(
+                    "Neighbor Operation:%s: neighbor_type:%s, project_id:%s, vpc_id:%s, network_type:%d, ip_index: %d,"
+                    "virtual_ip_address:%s, virtual_mac_address:%s, neighbor_host_ip_address:%s, tunnel_id:%d\n",
+                    aca_get_operation_string(current_NeighborState.operation_type()),
+                    aca_get_neighbor_type_string(
```
Comment (xieus) [inline src/dp_abstraction/aca_dataplane_ovs.cpp:437]: Is it possible to throw an exception in this long aca_get_neighbor_type_string?

**Review by xieus — COMMENTED** [review]
@er1cthe0ne Thanks for pushing the remaining fixes promptly. I left a few minor comments to the coding part. 

Regarding the openflow table doc, let us review it together today.

File: src/dp_abstraction/aca_dataplane_ovs.cpp (hunk 1 above)
Comment (er1cthe0ne) [reply src/dp_abstraction/aca_dataplane_ovs.cpp:217]: yes, that's a good suggestion, unfortunately the c++ [[deprecated]] keyword is for function level only.

File: src/dp_abstraction/aca_dataplane_ovs.cpp (hunk 2 above)
Comment (er1cthe0ne) [reply src/dp_abstraction/aca_dataplane_ovs.cpp:129]: got it. The current port creation code doesn't consume the virtual IP address(s). It would be consumed in security group code to allow traffic come out from the VM with certain virtual IPs and virtual mac (it also should look at the allow_address_pairs in PortConfiguration).

File: src/dp_abstraction/aca_dataplane_ovs.cpp (hunk 3 above)
Comment (er1cthe0ne) [reply src/dp_abstraction/aca_dataplane_ovs.cpp:372]: fixed

File: src/dp_abstraction/aca_dataplane_ovs.cpp (hunk 4 above)
Comment (er1cthe0ne) [reply src/dp_abstraction/aca_dataplane_ovs.cpp:377]: fixed, that's a good change.

File: src/dp_abstraction/aca_dataplane_ovs.cpp (hunk 5 above)
Comment (er1cthe0ne) [reply src/dp_abstraction/aca_dataplane_ovs.cpp:437]: no, it is just print error when the input is invalid :)

File: build/aca-machine-init.sh [hunk 6]
```diff
@@ -117,17 +117,12 @@ echo "7--- installing pulsar dependacies ---" && \
...
-ovs-vsctl add-br br-int -- add-br br-tun
-ovs-vsctl \
-    -- add-port br-int patch-tun \
-    -- set interface patch-tun type=patch options:peer=patch-int \
-    -- add-port br-tun patch-int \
-    -- set interface patch-int type=patch options:peer=patch-tun
-
-ovs-ofctl add-flow br-tun "table=0, priority=1,in_port="patch-int" actions=resubmit(,2)"
-ovs-ofctl add-flow br-tun "table=2, priority=0 actions=resubmit(,22)"
+if [ "$1" == "delete-bridges" ]; then
```
Comment (er1cthe0ne) [inline build/aca-machine-init.sh:120]: @cj-chung please take a look to confirm this approach is okay for our environment.

File: src/dp_abstraction/aca_dataplane_ovs.cpp (hunk 2 above)
Comment (xieus) [reply src/dp_abstraction/aca_dataplane_ovs.cpp:128]: Let me rephrase the question: if an interface has more than 1 ports, the current ACA would only process and store the first ip, and other ips are ignored. We will need to handle other ips too.

**Review by xieus — APPROVED** [review]
